lnsocket/ln.go

```diff
@@ -0,0 +1,280 @@
+	arr, err := base64.StdEncoding.DecodeString(key)
+	if err != nil {
+		return err
+	}
+
+	priv, _ := btcec.PrivKeyFromBytes(arr)
+	ln.PrivKeyECDH = &keychain.PrivKeyECDH{PrivKey: priv}
+
+	return nil
+}
+
+// Read reads from connection
+func (ln *LN) Read() (uint16, []byte, error) {
+	if ln.Conn == nil {
+		return 0, nil, io.ErrClosedPipe
+	}
+
+	res := make([]byte, 65535)
+	ln.Conn.SetDeadline(time.Now().Add(ln.Timeout))
+	n, err := ln.Conn.Read(res)
+	if err != nil {
+		return 0, nil, err
+	}
+	if n < 2 {
+		return 0, nil, fmt.Errorf("read too small")
+	}
+	res = res[:n]
+	msgtype := ParseMsgType(res)
+	return msgtype, res[2:], nil
+}
+
+// Write writes to connection
+func (ln *LN) Write(b []byte) (int, error) {
+	if ln.Conn == nil {
+		return 0, io.ErrClosedPipe
+	}
+	ln.Conn.SetDeadline(time.Now().Add(ln.Timeout))
+	return ln.Conn.Write(b)
+}
+
+// Handshake performs a handshake
+func (ln *LN) Handshake() error {
+	t, data, err := ln.Read()
+	typ := lnwire.MessageType(t)
+	if err != nil {
+		return err
+	}
+
+	if typ != lnwire.MsgInit {
+		return fmt.Errorf("unexpected message type: %v", typ)
+	}
+
+	init := lnwire.Init{}
+	init.Decode(bytes.NewReader(data), 0)
+
+	// Set some conservative options
+	result := GetMandatoryFeatures(&init)
+	features := ToRawFeatureVector(result)
+	if _, ok := result[0]; ok {
+		features = lnwire.NewRawFeatureVector(lnwire.DataLossProtectRequired)
+	} else {
+		if init.GlobalFeatures.IsEmpty() {
+			// Features should be ok for Eclair
+		} else {
+			// Old corelightning
+			features.Unset(lnwire.PaymentAddrRequired)
+			features.Unset(lnwire.PaymentAddrOptional)
+		}
+	}
+
+	// Override
+	/*
+		features = lnwire.NewRawFeatureVector()
+		features.Set(lnwire.WumboChannelsOptional)
+		features.Set(lnwire.DataLossProtectRequired)
+		features.Set(lnwire.DataLossProtectOptional)
+	*/
+
+	initReplyMsg := lnwire.NewInitMessage(features, features)
+
+	var b bytes.Buffer
+	_, err = lnwire.WriteMessage(&b, initReplyMsg, 0)
+	if err != nil {
+		return err
+
+	}
+
+	_, err = ln.Write(b.Bytes())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Ping sends a ping message
+func (ln *LN) Ping() error {
+	ping := lnwire.NewPing(16)
+	var b bytes.Buffer
+
+	_, err := lnwire.WriteMessage(&b, ping, 0)
+	if err != nil {
+		return err
+	}
+	_, err = ln.Write(b.Bytes())
+	if err != nil {
+		return err
+	}
+
+	for {
+		t, _, err := ln.Read()
+		typ := lnwire.MessageType(t)
+		if typ == lnwire.MsgPong {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+	}
+}
+
+func (ln *LN) connectWith(netAddr *lnwire.NetAddress) error {
+	var (
+		err  error
+		conn *brontide.Conn
+	)
+
+	if ln.Proxy == nil {
+		conn, err = brontide.Dial(ln.PrivKeyECDH, netAddr, ln.Timeout, net.DialTimeout)
+	} else {
+		conn, err = brontide.Dial(ln.PrivKeyECDH, netAddr, ln.Timeout, ln.Proxy.Dial)
+	}
+
+	ln.Conn = conn
+
+	return err
+}
+
+// Connect connects to pubkey@host
+func (ln *LN) Connect(s string) error {
```

Comment:
Connect function clearly does not do just one thing. Move string parsing in another function